File: android/src/main/java/expo/modules/mediastore/MediaStorePermissions.kt
package expo.modules.mediastore

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.ContextCompat

class MediaStorePermissions(private val context: Context) {
  fun checkStatus(): Map<String, Any?> {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      return mapOf(
        "granted" to isPermissionGranted(Manifest.permission.READ_MEDIA_AUDIO) &&
          isPermissionGranted(Manifest.permission.READ_MEDIA_VIDEO) &&
          isPermissionGranted(Manifest.permission.READ_MEDIA_IMAGES),
        "audio" to isPermissionGranted(Manifest.permission.READ_MEDIA_AUDIO),
        "video" to isPermissionGranted(Manifest.permission.READ_MEDIA_VIDEO),
        "images" to isPermissionGranted(Manifest.permission.READ_MEDIA_IMAGES)
      )
    } else {
      return mapOf(
        "granted" to isPermissionGranted(Manifest.permission.READ_EXTERNAL_STORAGE),
        "audio" to isPermissionGranted(Manifest.permission.READ_EXTERNAL_STORAGE),
        "video" to isPermissionGranted(Manifest.permission.READ_EXTERNAL_STORAGE),
        "images" to isPermissionGranted(Manifest.permission.READ_EXTERNAL_STORAGE)
      )
    }
  }

  suspend fun request(): Map<String, Any?> {
    return checkStatus()
  }

  private fun isPermissionGranted(permission: String): Boolean {
    return ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED
  }
}

File: ios/MediaStorePermissions.swift
import Photos

class MediaStorePermissions {
  func checkStatus() -> [String: Any?] {
    let status = PHPhotoLibrary.authorizationStatus(for: .readWrite)

    return [
      "granted": status == .authorized || status == .limited,
      "audio": status == .authorized || status == .limited,
      "video": status == .authorized || status == .limited,
      "images": status == .authorized || status == .limited
    ]
  }

  func request() async -> [String: Any?] {
    let currentStatus = PHPhotoLibrary.authorizationStatus(for: .readWrite)

    if currentStatus == .authorized || currentStatus == .limited {
      return checkStatus()
    }

    let status = await PHPhotoLibrary.requestAuthorization(for: .readWrite)

    return [
      "granted": status == .authorized || status == .limited,
      "audio": status == .authorized || status == .limited,
      "video": status == .authorized || status == .limited,
      "images": status == .authorized || status == .limited
    ]
  }
}
